source/extensions/transport_sockets/tls/context_config_impl.cc

```diff
@@ -392,7 +392,15 @@ ServerContextConfigImpl::ServerContextConfigImpl(
           PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, require_client_certificate, false)),
       ocsp_staple_policy_(ocspStaplePolicyFromProto(config.ocsp_staple_policy())),
       session_ticket_keys_provider_(getTlsSessionTicketKeysConfigProvider(factory_context, config)),
-      disable_stateless_session_resumption_(getStatelessSessionResumptionDisabled(config)) {
+      disable_stateless_session_resumption_(getStatelessSessionResumptionDisabled(config)),
+      full_scan_certs_on_sni_mismatch_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, full_scan_certs_on_sni_mismatch, false)) {
+
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.allow_full_scan_certs_on_sni_mismatch") &&
+      !config.has_full_scan_certs_on_sni_mismatch()) {
+    full_scan_certs_on_sni_mismatch_ = true;
+  }
```

Comment:
I think I would invert this logic and have the runtime guard lead this to getting set to false. So rename to something like `envoy_reloadable_features_no_full_scan_certs_on_sni_mismatch` or something like that. This will match the config setting which is false by default.

Reply:
@mattklein123 
Per my understanding, if we just want the config setting is false by default as you mentioned, 

> b) The default for the configuration would be false (abort and do not do a full scan)

we don't need RUNTIME GUARD because we have
https://github.com/envoyproxy/envoy/blob/b12f94357a2d2d885bd6ceb07db388df48d93b28/source/extensions/transport_sockets/tls/context_config_impl.cc#L396-L397

If we want to test if people run into trouble with full scan, we need RUNTIME_GUARD to force enabling full scan if the config is not set explicitly. And if no issues raised for a long time, we can remove this RUNTIME_GUARD.

Please correct me if I misunderstood.

Reply:
This logic is saying: if <runtime guard enabled> && !<field is set>, set full scan to true. The first clause defaults to true. The 2nd clause will generally evaluate to true on upgrade. Thus, on upgrade this will be true and we will do a full scan. I think on upgrade we want the default to be the same as the config default which is false?

Reply:
Yes, but we can set the default by [PROTOBUF_GET_WRAPPED_OR_DEFAULT](https://github.com/envoyproxy/envoy/blob/dede92e6e5888cdf27320bf3b6bc9987657d6e22/source/common/protobuf/utility.h#L22), why do we need RUNTIME GUARD here? There is no difference we introducing RUNTIME GUARD or not.

You might want this logic 
```c++
  if (Runtime::runtimeFeatureEnabled(
          "envoy.reloadable_features.allow_full_scan_certs_on_sni_mismatch") &&
      !config.has_full_scan_certs_on_sni_mismatch()) {
    full_scan_certs_on_sni_mismatch_ = false;
  }
```
But the default value is already set by below, there is no need to check RUNTIME GUARD and set it again
```c++
 full_scan_certs_on_sni_mismatch_( 
     PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, full_scan_certs_on_sni_mismatch, false))
```

Reply:
I was always thinking about "disable full scan by default" because there is DoS issue when enabling full scan, and I was not aware of that it is indeed possible that users might meet error when they disable full scan. So I was stuck, I never thought we need to set back to full scan via runtime flag.

@soulxu Thanks for your comments. It helps a lot.